### ISSUE TITLE
Sparse Latticed Tensors (SLT)

### ISSUE THREAD

**Issue by PierreQuinton** [albanD/subclass_zoo#54]
## Sparse Latticed Tensors
With my research partner @ValerianRey, we're in the process of creating a new type of Tensor to represent sparsity with a lattice pattern. Our implementation is currently to represent these tensors with these attributes:
- A dense tensor called `physical` that contains the actual data.
- A `basis` matrix (the basis of the lattice) and an `offset`, representing the affine transformation from an index in the `physical` tensor to an index in the virtual (i.e. the sparse) tensor: `virtual_index = basis @ physical_index + offset`. The shape of `basis` is thus `[virtual.ndim, physical.ndim]`.
- The `shape` of the virtual tensor.

The non-zero elements can only be located on the affine lattice defined by `basis` and `offset` and this is why we call this a Sparse Latticed Tensor.
We will see that this structure can be very efficient in terms of memory, but also in terms of operations: Many of them can be translated as operations on the (smaller) physical tensor.

---

### Examples

<details>
<summary>Diagonal matrix</summary>

virtual:
```
[[1 0 0]
 [0 2 0]
 [0 0 3]]
```

`physical=tensor([1, 2, 3])`, `basis=[[1], [1]]`, `offset=[0, 0]`, `shape=[3, 3]`
Moving by 1 in the physical tensor (e.g. going from 1 to 2) makes you move by 1 line and 1 column in the virtual tensor. You can thus think of the basis as some form of strides.
</details>

<details>
<summary>Anti-diagonal matrix</summary>

virtual:
```
[[0 0 1]
 [0 2 0]
 [3 0 0]]
```
`physical=tensor([1, 2, 3])`, `basis=[[1], [-1]]`, `offset=[0, 2]`, `shape=[3, 3]`. Notice the negative value in the basis.
</details>

<details>
<summary>Block-diagonal matrix</summary>

virtual:
```
[[a b 0 0 0 0]
 [c d 0 0 0 0]
 [0 0 e f 0 0]
 [0 0 g h 0 0]
 [0 0 0 0 i j]
 [0 0 0 0 k l]]
```
`physical=tensor([[[a, b], [c, d]], [[e, f], [g, h]], [[i, j], [k, l]]])`, `basis=[[2, 1, 0], [2, 0, 1]]`, `offset=[0, 0]`, `shape=[6, 6]`
</details>

<details>
<summary>Flattened diagonal matrix</summary>

virtual:
```
[1 0 0 0 2 0 0 0 3]
```
`physical=tensor([1, 2, 3])`, `basis=[[4]]`, `offset=[0]`, `shape=[9]`
</details>

<details>
<summary>Diagonal tensor of higher ndim</summary>

virtual:
```
[[[1 0 0] [0 0 0] [0 0 0]]
 [[0 0 0] [0 2 0] [0 0 0]]
 [[0 0 0] [0 0 0] [0 0 3]]]
```
`physical=tensor([1, 2, 3])`, `basis=[[1], [1], [1]]`, `offset=[0, 0, 0]`, `shape=[3, 3, 3]`
</details>

<details>
<summary>Circulant matrix (e.g. expanded conv1d kernel)</summary>

virtual:
```
[[1 2 3 0 0 0]
 [0 1 2 3 0 0]
 [0 0 1 2 3 0]
 [0 0 0 1 2 3]]
```
`physical=tensor([1, 2, 3]).expand(4, 3)`, `basis=[[1, 0], [1, 1]]`, `offset=[0, 0]`, `shape=[4, 6]`
Note that here, only 3 values are stored in memory, because expand efficiently supports repetition.
</details>

<details>
<summary>Padding</summary>

virtual:
```
[[0 0 0 0]
 [1 2 0 0]
 [3 4 0 0]]
```
`physical=tensor([[1, 2], [3, 4]])`, `basis=[[1, 0], [0, 1]]` (identity), `offset=[1, 0]`, `shape=[3, 4]`.
The offset and shape give a lot of control over the type of padding we want to have.
</details>

<details>
<summary>Bonus: cross shape (🇨🇭)</summary>

virtual:
```
[[1 1 0 0 1 1]
 [1 1 0 0 1 1]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [1 1 0 0 1 1]
 [1 1 0 0 1 1]]
```
`physical=ones(2, 2, 2, 2)`, `basis=[[0, 1, 0, 4], [1, 0, 4, 0]]`, `offset=[0, 0]`, `shape=[6, 6]`
</details>

<details>
<summary>Bonus: checkerboard (🏁)</summary>

virtual:
```
[[1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]]
```
`physical=ones(4, 4, 2)`, `basis=[[0, 2, 1], [2, 0, 1]]`, `offset=[0, 0]`, `shape=[8, 8]`
</details>

---

### Interactive plotter

You can play with Sparse Latticed Tensor on Geogebra at https://www.geogebra.org/calculator/uvdawrwc (we advise to use the keyboard arrows to move things by exactly one unit at a time).

### Use-cases
To me, this representation for sparse tensors seems very powerful, because operations like matrix multiplication (or more generally any einsum) can be done on the physical tensors directly, as long as a few things are being taken care of. Here is a list of potential use-cases:

<details>
<summary>Efficient computation of per-sample gradients</summary>

When using a model that treats each element of the batch independently, the Jacobian of the per-sample losses with respect to any batch of activations is block-diagonal. Therefore, rather than representing the Jacobian (i.e. the batched grads) as a normal tensor (as is done in the [per-sample gradients tutorial](https://docs.pytorch.org/tutorials/intermediate/per_sample_grads.html#per-sample-grads-the-efficient-way-using-function-transforms)) we could store it on a sparse latticed tensor. All Jacobians with respect to internal activations will also be block-diagonal (with blocks of 1 row), so the sparsity is preserved through most of the backward pass (the only Jacobians that are dense are those with respect to the model parameters). So this would save tons of memory (up to a division by batch_size). This is something we would love to have to make https://github.com/SimplexLab/TorchJD much more efficient, and this is the initial reason we started working on that.
</details>

<details>
<summary>Efficient computations following `torch.diag` / `torch.eye`</summary>

When a user calls `torch.diag` (or `torch.eye`), the result could be a sparse latticed tensor instead of a dense tensor. This type would be preserved for as long as the operations applied to this tensor yield a sparse tensor with a lattice structure. However, whenever the result would be an actually dense tensor, the operation would just return a normal Tensor.

For instance:
```python
data = tensor([1., 2., 3.])

# Create the sparse latticed tensor using torch.diag
t1 = torch.diag(data, layout=torch.sparse_latticed)
# [[1 0 0] 
#  [0 2 0]
#  [0 0 3]]
# physical=tensor([1., 2., 3.]), basis=[[1], [1]], offset=[0, 0], shape=[3, 3]

# Make an operation that preserves sparsity (zero-preserving pointwise operation)
t2 = torch.square(t1) 
# [[1 0 0] 
#  [0 4 0]
#  [0 0 9]]
# physical=tensor([1., 4., 9.]), basis=[[1], [1]], offset=[0, 0], shape=[3, 3]

# Make another operation that changes the shape but still preserves sparsity
t3 = torch.cat([t1, t2], dim=1)
# [[1 0 0 1 0 0] 
#  [0 2 0 0 4 0]
#  [0 0 3 0 0 9]]
# physical=tensor([[1., 2., 3.], [1., 4., 9.]]), basis=[[0, 1], [3, 1]], offset=[0, 0], shape=[3, 6]

# Make an operation that does not preserve sparsity
t4 = torch.sum(t3, dim=0)
# [1 2 3 1 4 9]
# t4 is a standard Tensor
```
</details>

<details>
<summary>Repeated output indices in `torch.einsum`</summary>

When doing einsum, we cannot repeat indices in the output (e.g. "ij, jk → iik"), but the mathematical sum defining those is actually well-defined, and the output would naturally be a SLT. We simply need to remove duplicates, and then provide the corresponding basis (e.g. `SLT(physical=einsum(..., "ij, jk -> ik"), basis=[[1, 0], [1, 0], [0, 1]]`).

</details>

<details>
<summary>Alternative implementation of convolutions</summary>

As seen in the example of the circulant matrix, a conv1d operation with no channel dimension can be represented by a multiplication with a circulant matrix, that is representable efficiently using a SparseLatticedTensor. More generally, any convNd operation with any number of input channel and output channel can be implemented as a simple tensordot with a very sparse tensor, that is representable as a SparseLatticedTensor.

Obviously, convNd operations are already super efficiently implemented, so there's no point in changing their implementation, but I just think it's cool to see that the SparseLatticedTensor has the potential to make the efficient implementation of convolutions trivial.
</details>

<details>
<summary>Neural network sparsification</summary>

Nvidia has done some experiments on 2:4 semi-structured sparsity: https://developer.nvidia.com/blog/structured-sparsity-in-the-nvidia-ampere-architecture-and-applications-in-search-engines/. I think that SparseLatticedTensors would allow similar experimentation, but without the cost of having to make custom cuda kernels for that.
</details>

<details>
<summary>Memory-efficient `torch.repeat`</summary>

When doing [torch.repeat](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.repeat.html), memory has to be copied as many times as the dimension is repeated. We could instead do a `torch.expand` on the physical tensor and use the basis to indicate which dimensions are repeated.

Standard use of torch.repeat:
```python
>>> x = torch.tensor([1, 2, 3])
>>> x.repeat(4, 2)
tensor([[ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3]])
```
Memory-efficient alternative using SparseLatticedTensor:
```python
>>> x = torch.tensor([1, 2, 3])
>>> physical = x.expand(4, 2, 3)
>>> basis = torch.tensor([[1, 0, 0], [0, 3, 1]])
>>> repeated = SparseLatticedTensor(physical, basis)  # No offset, shape can be inferred
>>> repeated.to_dense()
tensor([[ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3],
        [ 1,  2,  3,  1,  2,  3]])
```
Since this one is not trivial, I'll explain it in more details:
- `physical` only stores 3 values in memory, because `torch.expand` produces a tensor that has a stride of 0 on the expanded dimensions. It contains all the values that we need, but the shape is not the one we want (it has shape [4, 2, 3] and we want shape [4, 6]). This is not something that we can fix simply by using tensor strides.
- the `basis` means that moving on the first dimension of the physical makes you move by 1 row and 0 column in the virtual tensor. Moving in the second dimension of the physical tensor makes you move by 0 row and 3 columns, and moving in the third dimension makes you move by 0 row and 1 column. If this seems confusing, try looking at the transpose of `basis`. 

</details>

---

### Operations

SLTs are compositional, meaning that operations on them can typically be translated into operations on the physical tensor, and often result in another SLT. We want the SLT structure to be closed under as many operations as possible.

Like for all other sparse tensors, every operation `f` applied element-wise and such that `f(0.0)=0.0` can be applied to the physical data only, and preserves the SLT structure.

`movedim` can be done easily by permuting the shape, offset and basis accordingly and no change is required on the physical.

If we elementwise multiply two SLTs, then the non-zero elements can only be at the lattice intersection of the two lattices defined by the SLTs. Similarly, if we elementwise add two SLTs, the resulting lattice should be the lattice sum of those defined by the SLTs.
The lattice intersection and sum can be characterized by selecting the basis.
This will probably rely on some appropriately selected normal form. The typical [HNF](https://en.wikipedia.org/wiki/Hermite_normal_form) will not be the best choice as it will not factorize lattices efficiently (this will need to be carefully designed, but can be done independently of the rest as long as the lattice intersection/sum abstraction exist).

The most meaningful operation example is the einsum, because it plays an important role everywhere. An einsum involving SLTs can be translated to an einsum on the physicals, and then the result is (sometimes) wrapped into a SLT. Note that a dense tensor can be wrapped as a SLT whose basis is the identity matrix.
The idea is to translate virtual indices into physical indices and see if some match.
The simplest example is the matrix multiplication of the identity with a matrix.
The identity matrix is defined as `I=SLT(ones(m), basis=[[1], [1]], offset=[0, 0], shape=[m, m])`. Multiplying this by a matrix `M` is `einsum(I, M, "ij, jk -> ik")`. The idea is to create indices corresponding to physical dimension. Here, one (`l`) for `I` and two (`m` and `n`) for `M`. Then we map the indices via the basis to the virtual indices: `l->i`, `l->j`, `m->j` and `n->k`. Finally, we group physical indices that can map to the same virtual indices. Here `{l, m}` for a group (they can map to `j`), and `{n}` for another (it can map to `k`). We pick a representative of each group and formulate the einsum on the physical: `einsum(ones(m), M, "l, lm -> lm")`. Note that translating the output indices from virtual to physical may result in some repetition of the output indices. We would then need to remove duplicates and wrap the result into an SLT that has an extra column in its basis for each duplicate that was removed.

Another example is the matrix multiplication of `SLT(physical=ones(6), basis=[[2], [2]], offset=[0, 0], shape=[12, 12])` and `SLT(physical=ones(4), basis=[[3], [3]], offset=[0, 0], shape=[12, 12])`. The result should be `SLT(physical=ones(2), basis=[[6], [6]], offset=[0, 0], shape=[12, 12])`. This means that this will require some amount of lattice intersection if done right.

### Progress and way forward

So far, we have a quite messy pull request in TorchJD, that adds some implementations for some of the most important functions on SLTs. Some of them work, some do not (yet). Here is the link in case you wanna take a look: https://github.com/SimplexLab/TorchJD/pull/484

What do you think would be the best way forward?

@albanD FYI This is the thing we started talking about at PyTorch conference Europe on Tuesday!
